vehicle/mercedes/api.go

```diff
@@ -0,0 +1,204 @@
+package mercedes
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"github.com/evcc-io/evcc/util/transport"
+	protos "github.com/evcc-io/evcc/vehicle/mercedes/pb"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+	"google.golang.org/protobuf/proto"
+)
+
+// API is an api.Vehicle implementation for Mercedes-Benz cars
+type API struct {
+	identity *Identity
+	log      *util.Logger
+	*request.Helper
+}
+
+type SetupAPI struct {
+	log     *util.Logger
+	account string
+	region  string
+	*request.Helper
+}
+
+func NewAPI(log *util.Logger, identity *Identity) *API {
+	client := request.NewHelper(log)
+
+	client.Transport = &transport.Decorator{
+		Base: &oauth2.Transport{
+			Source: identity,
+			Base:   client.Transport,
+		},
+		Decorator: transport.DecorateHeaders(mbheaders(false, identity.region)),
+	}
+
+	return &API{
+		Helper:   client,
+		identity: identity,
+		log:      log,
+	}
+}
+
+func NewSetupAPI(log *util.Logger, account string, region string) *SetupAPI {
+	client := request.NewHelper(log)
+
+	client.Transport = &transport.Decorator{
+		Base:      client.Transport,
+		Decorator: transport.DecorateHeaders(mbheaders(true, region)),
+	}
+
+	return &SetupAPI{
+		Helper:  client,
+		log:     log,
+		region:  region,
+		account: account,
+	}
+}
+
+func (v *API) Vehicles() ([]string, error) {
+	var res VehiclesResponse
+
+	url := fmt.Sprintf("%s/v2/vehicles", getBffUri(v.identity.region))
+
+	err := v.GetJSON(url, &res)
+
+	var vehicles []string
+	for _, v := range res.assignedVehicles {
+		vehicles = append(vehicles, v.fin)
+	}
+
+	return vehicles, err
+}
+
+func (v *API) Status(vin string) (StatusResponse, error) {
+	var res StatusResponse
+
+	uri := fmt.Sprintf("%s/v1/vehicle/%s/vehicleattributes", getWidgetUri(v.identity.region), vin)
+
+	data, err := v.GetBody(uri)
+
+	if err == nil {
+		message := &protos.VEPUpdate{}
+		err := proto.Unmarshal(data, message)
+
+		if err == nil {
+			if val, ok := message.Attributes["odo"]; ok {
+				if val.GetNilValue() {
+					res.VehicleInfo.Odometer.Value = -1
```

Comment:
die -1 sind nicht schön. Lieber nil pointer für nicht gelieferten Wert?

Reply:
Auf 0 geändert. Der Provider kann ja kein nil zurückgeben und in den anderen vehicles wird in dem case auch 0 zurückgegeben.